<a href="https://colab.research.google.com/github/BridgingAISocietySummerSchools/Data-Science-AI-Python-Course/blob/main/01_introduction/02_data_structures.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# 📦 Data Structures — Lists, Tuples and Dictionaries

> **Track:** Introduction Session — Notebook 2 of 5 · **Estimated time:** ~18 min · **Difficulty:** complete beginner

In Notebook 1 each variable held exactly one value. That is fine for a price or an age — and hopeless for a survey with 300 answers. This notebook introduces Python's three everyday **containers**, the boxes that hold *many* values at once. Everything you analyse later today sits in one of them.

## 🎯 Learning objectives

By the end of this notebook you will be able to:

1. Explain **why** collections are needed, and pick the right one for a job.
2. Use a **list**: create it, read items by position (including from the end), measure it with `len()`, change an item, and grow it with `.append()`.
3. Use a **tuple**, say what "immutable" means, and **unpack** one into separate variables.
4. Use a **dictionary**: read and write values by their **key**, test membership with `in`, and look at `.keys()` / `.values()`.
5. Recognise a **list of dictionaries** as a table — the shape pandas takes over in Notebook 5.

## ✅ Prerequisites

Notebook 1 ([`01_welcome_to_python.ipynb`](01_welcome_to_python.ipynb)): running cells, variables, the types `int` / `float` / `str` / `bool`, arithmetic, and f-strings (including `:.2f`).


## 1. Why data structures? One value per box does not scale

Suppose you collect a feedback score (1–5) from every participant in this room. With Notebook 1's tools you would write:

```python
score_1 = 4
score_2 = 5
score_3 = 3
score_4 = 5
score_5 = 4      # ... and now imagine 300 participants
```

Three things are already broken: you cannot compute an average without naming all 300 variables by hand, you cannot add a 301st participant without editing your code, and nothing in the code says that these values belong together.

Real data never arrives as single values. It arrives in **collections** — a column of measurements, a row describing one person, a table of both. Python gives you three everyday collections, and this table is the map for the rest of the notebook:

|  | Looks like | You look things up by | Can it change? | Use it for |
|---|---|---|---|---|
| **List** | `[4, 5, 3]` | **position**: `scores[0]` | yes — items change, list grows | a *series* of similar things: measurements, answers, rows |
| **Tuple** | `(52.52, 13.40)` | **position**: `location[0]` | no — fixed once created | a small *fixed record*: coordinates, a pair, a function's several results |
| **Dictionary** | `{"age": 31}` | **name (key)**: `person["age"]` | yes | *one thing* described by labelled fields |

One sentence to carry away: **lists number their contents, dictionaries name them, tuples refuse to be edited.** Come back to this table whenever you are unsure which one you want.


## 2. Lists — many values in numbered compartments

A **list** is one box with numbered compartments. You build it with square brackets and commas:

```
scores  =  [  4  ,  5  ,  3  ,  5  ,  4  ]
position:     0     1     2     3     4
```

A compartment's position number is called its **index**, and you use it to pull one value out: `scores[2]`. One convention trips up absolutely everybody at first: **Python starts counting at 0.** The first item is item `0`, the second is item `1`. So the *fifth* item of our list is `scores[4]`.

In [1]:
scores = [4, 5, 3, 5, 4]          # feedback scores from five participants

print(f"the whole list   : {scores}")
print(f"first item [0]   : {scores[0]}")      # counting starts at 0!
print(f"third item [2]   : {scores[2]}")
print(f"last item [-1]   : {scores[-1]}")     # a negative index counts from the END
print(f"how many items   : {len(scores)}")
print(f"average score    : {sum(scores) / len(scores):.2f}")

the whole list   : [4, 5, 3, 5, 4]
first item [0]   : 4
third item [2]   : 3
last item [-1]   : 4
how many items   : 5
average score    : 4.20



**Reading the output:** `scores[0]` gives `4`, the *first* score, and `scores[2]` gives `3`, the *third*. The negative index is a lovely shortcut: `scores[-1]` means "the last one" without your having to know how long the list is (`scores[-2]` is the second-to-last). `len(scores)` reports `5` — the number of items, which is also the number of survey responses.

The last line is the real payoff. `sum(scores) / len(scores)` is the average — `4.20` — and that single line works unchanged for 5 scores or 5,000. That is what a collection buys you: **one name for the whole dataset.**

In [2]:
# 1) Change one item — participant 3 corrected their score afterwards:
scores[2] = 4
print(f"after the correction : {scores}")

# 2) Grow the list — a late participant hands in their form:
scores.append(5)
print(f"after .append(5)     : {scores}")
print(f"number of scores now : {len(scores)}")

# 3) Slicing, as a one-line taster — the first three scores:
print(f"the first three      : {scores[0:3]}")

after the correction : [4, 5, 4, 5, 4]
after .append(5)     : [4, 5, 4, 5, 4, 5]
number of scores now : 6
the first three      : [4, 5, 4]



**Reading the output:** the third item really did change from `3` to `4`, and after `.append(5)` the list is one item longer — `len()` now reports `6`. Lists are **mutable**: the box itself is modified, so every variable pointing at that list sees the new contents. (A small live-session warning: run this cell twice and `.append()` runs twice, so the list keeps growing. If a number ever looks wrong, use **Runtime → Restart and run all** in Colab.)

Line 3 is your first **slice**: `scores[0:3]` means "from position 0 up to *but not including* position 3" — three items. That asymmetry has a logic to it, and Track 2's [`03_lists_data_structures.ipynb`](../02_advanced_self_learning/03_lists_data_structures.ipynb) gives you the full model. For today, read `scores[0:3]` as "the first three" and move on.


### 🎯 Try it (1–2 min) — Build a list and poke at it

In the cell below, create a list called `cities` holding **four** city names (any cities — text goes in quotes). Then:

1. print the first city,
2. print the last city using a **negative** index,
3. replace the second city with a different one and print the whole list,
4. append one more city and print the new `len()`.

In [3]:
# Your turn 👇  Fill in the list, then press Shift + Enter.

cities = ["...", "...", "...", "..."]

# 1. print the first city
# 2. print the last city with a negative index
# 3. replace the second city, then print the list
# 4. append a city and print len(cities)


<details><summary>💡 Click to reveal the solution</summary>

```python
cities = ["Berlin", "Nairobi", "Bogotá", "Jakarta"]

print(cities[0])        # Berlin
print(cities[-1])       # Jakarta

cities[1] = "Lagos"
print(cities)           # ['Berlin', 'Lagos', 'Bogotá', 'Jakarta']

cities.append("Oslo")
print(len(cities))      # 5
```

**Reasoning.** `cities[0]` is the first element because counting starts at 0, and `cities[-1]` walks in from the other end, which is why it is the safer way to say "the last one". Assigning `cities[1] = "Lagos"` overwrites that one compartment and leaves the rest untouched — the list keeps its length. `.append()` is different in kind: it makes the list *longer*, which is why `len()` goes from 4 to 5.
</details>


## 3. Tuples — a small record that must not change

A list is a shopping list: things get added, crossed off, rewritten. A **tuple** is more like a printed label or a stamped coordinate — created once, then fixed. You build one with **round** brackets:

```python
location = (52.52, 13.40)     # (latitude, longitude) of Berlin
```

You read items exactly as in a list (`location[0]`), and `len()` works too. What you *cannot* do is change an item. That sounds like a limitation; it is actually the point. Some things belong together and must never drift apart — the two halves of a coordinate, the three parts of a date — and a tuple guarantees that nobody edits one half by accident.

In [4]:
location = (52.52, 13.40)          # round brackets → a tuple

print(f"the whole tuple : {location}")
print(f"latitude  [0]   : {location[0]}")
print(f"longitude [1]   : {location[1]}")
print(f"how many parts  : {len(location)}")

# Unpacking: hand each part its own variable name, in one line
lat, lon = location
print(f"unpacked        : lat = {lat}, lon = {lon}")

# A tuple refuses to be edited. Uncomment the next line, run it, and read the error:
# location[0] = 48.14

the whole tuple : (52.52, 13.4)
latitude  [0]   : 52.52
longitude [1]   : 13.4
how many parts  : 2
unpacked        : lat = 52.52, lon = 13.4



**Reading the output:** indexing and `len()` behave just like a list. (Python prints `13.4`, not `13.40` — numbers don't remember how you typed them, exactly as in Notebook 1.) The interesting line is `lat, lon = location` — **unpacking**. Python matches the variables on the left to the items on the right, in order, so you get two clearly named values (`lat`, `lon`) instead of having to remember that latitude was "the one at position 0". Unpacking is the reason tuples feel pleasant in practice.

If you uncomment the last line, Python stops you:

```text
TypeError: 'tuple' object does not support item assignment
```

Read it as *"you may not assign into a tuple"*. That is the meaning of **immutable**. So when is a tuple the right choice?

- **Fixed records** — a coordinate `(lat, lon)`, an RGB colour `(31, 119, 180)`, a date `(2026, 8, 22)`.
- **Values that belong together** and should never be edited separately.
- **Returning several results from a function** — you will see `return x, y` in Notebook 4; Python hands back a tuple.
- Anywhere you want the *protection*: if a value must not change, make it impossible rather than remembering not to.

Rule of thumb: a **list** is a series of *many similar* things; a **tuple** is *one thing with a fixed number of parts*.


### 🎯 Try it (1–2 min) — Unpack a record

The tuple below describes one participant: name, age, field of study. **Unpack** it into three variables and print one sentence with an f-string, for example `Amina is 31 and works in Sociology.`

Bonus (think, don't run): what happens if you write `participant[1] = 32`?

In [5]:
# Your turn 👇
participant = ("Amina", 31, "Sociology")

# 1. unpack into name, age, field
# 2. print(f"...")


<details><summary>💡 Click to reveal the solution</summary>

```python
participant = ("Amina", 31, "Sociology")

name, age, field = participant
print(f"{name} is {age} and works in {field}.")
# Amina is 31 and works in Sociology.
```

**Reasoning.** The three names on the left are matched to the three items on the right *by position*, so the order matters — swap `age` and `field` and Python will happily give you nonsense, because it checks the count, not the meaning. (The count it does check: two names for a three-item tuple raises `ValueError: too many values to unpack`.)

The bonus: `participant[1] = 32` fails with `TypeError: 'tuple' object does not support item assignment`. To "change" a tuple you build a new one — `participant = (name, 32, field)` — which is exactly the honesty a tuple is there to force.
</details>


## 4. Dictionaries — look things up by name, not by number

Look again at `participant[1]`. Was position 1 the age or the field of study? With a list or tuple *you* have to remember the meaning of every position — and in real projects, you won't.

A **dictionary** fixes that by labelling each value. You write it with **curly** braces as `key: value` pairs, and you look things up by the label — the **key** — instead of a number:

```
        list / tuple                        dictionary
   position 0 → "Amina"              "name"  → "Amina"
   position 1 → 31                   "age"   → 31
   position 2 → "Sociology"          "field" → "Sociology"
```

Think of a paper form, or an index card: each field has a printed name, and you read the box next to the name you want. The order of the fields does not matter; their names do.

In [6]:
participant = {
    "name": "Amina",
    "age": 31,
    "field": "Sociology",
    "attended_before": False,
}

print(f"the whole record : {participant}")
print(f"one value by key : {participant['age']}")
print(f"how many fields  : {len(participant)}")
print(f"{participant['name']} is {participant['age']} and works in {participant['field']}.")

the whole record : {'name': 'Amina', 'age': 31, 'field': 'Sociology', 'attended_before': False}
one value by key : 31
how many fields  : 4
Amina is 31 and works in Sociology.



**Reading the output:** `participant["age"]` reads almost like English — "the age of this participant" — and, unlike `participant[1]`, it will still make sense to you next month. Note that each field keeps its own type: `"Amina"` is a `str`, `31` an `int`, `False` a `bool`. A dictionary happily mixes types, because its job is to describe *one thing* from several angles. `len()` on a dictionary counts the **fields**, not the characters.

(One small syntax detail: inside an f-string written with double quotes, the key needs *single* quotes — `f"{participant['age']}"`. Outside f-strings, `participant["age"]` with double quotes is the usual style.)

Records are rarely finished when you first write them. Let's add a field, correct one, and ask what the record contains:

In [7]:
# Add a NEW field — just assign to a key that isn't there yet:
participant["country"] = "Kenya"

# Update an EXISTING field — exactly the same syntax:
participant["age"] = 32

print(f"the record now : {participant}")
print(f"the labels     : {participant.keys()}")
print(f"the contents   : {participant.values()}")

# Membership: does this record even have such a field?
print(f'"country" in participant -> {"country" in participant}')
print(f'"email"   in participant -> {"email" in participant}')

# The safe way to read a field that might be missing:
print(f'get("email")              -> {participant.get("email")}')
print(f'get("email", "not given") -> {participant.get("email", "not given")}')

# participant["email"]    # ← uncomment, run, and read the KeyError

the record now : {'name': 'Amina', 'age': 32, 'field': 'Sociology', 'attended_before': False, 'country': 'Kenya'}
the labels     : dict_keys(['name', 'age', 'field', 'attended_before', 'country'])
the contents   : dict_values(['Amina', 32, 'Sociology', False, 'Kenya'])
"country" in participant -> True
"email"   in participant -> False
get("email")              -> None
get("email", "not given") -> not given



**Reading the output:** adding and updating look identical — `participant["country"] = "Kenya"` created a new field because no such key existed, while `participant["age"] = 32` overwrote the old value. `.keys()` gives you the field names and `.values()` their contents; Python shows them wrapped as `dict_keys([...])` and `dict_values([...])`, which you can simply read as "the labels" and "the contents" of the record.

`in` answers a question you will ask constantly with real data: *is this field present at all?* `"country" in participant` is `True`, `"email" in participant` is `False`. It matters, because reading a key that isn't there does not give you an empty value — it stops your program:

```text
KeyError: 'email'
```

That is a `KeyError`, one of the most common errors in real data work, since real records have missing fields. The one-line cure is `.get()`: `participant.get("email")` returns `None` instead of crashing, and `participant.get("email", "not given")` lets you choose the fallback yourself. Use `[...]` when the field must exist, `.get(...)` when it might not.


### 🎯 Try it (1–2 min) — A record for this session

Build a dictionary called `session` with the keys `"title"` (any text), `"minutes"` (a number) and `"level"` (`"beginner"`). Then:

1. print the value stored under `"minutes"`,
2. add a new field `"track"` with the value `"introduction"`,
3. print whether `"price"` is a field of `session`,
4. print `session.keys()`.

In [8]:
# Your turn 👇
session = {
    "title": "...",
    "minutes": 0,
    "level": "beginner",
}

# 1. print the minutes
# 2. add a "track" field
# 3. is "price" a field?
# 4. print the keys


<details><summary>💡 Click to reveal the solution</summary>

```python
session = {
    "title": "Data Structures",
    "minutes": 18,
    "level": "beginner",
}

print(session["minutes"])        # 18
session["track"] = "introduction"
print("price" in session)        # False
print(session.keys())            # dict_keys(['title', 'minutes', 'level', 'track'])
```

**Reasoning.** Reading uses the key in square brackets, and writing to a key that does not exist yet *creates* the field — which is why step 2 needs no special "add" command. `"price" in session` is `False`, and that check is exactly what saves you from a `KeyError` a moment later. Notice the new `"track"` key appears at the *end* of `.keys()`: dictionaries remember insertion order, but you should never rely on position — that is what keys are for.
</details>


## 5. The bridge — a list of dictionaries is a table

Now combine two of the three. One dictionary describes one participant; a **list of dictionaries** describes *all* of them — one dictionary per row.

In [9]:
participants = [
    {"name": "Amina", "age": 32, "field": "Sociology"},
    {"name": "Ben",   "age": 27, "field": "Law"},
    {"name": "Chen",  "age": 41, "field": "Medicine"},
]

print(f"rows (participants) : {len(participants)}")
print(f"the second row      : {participants[1]}")
print(f"one single cell     : {participants[1]['name']}")   # row 1, column "name"

rows (participants) : 3
the second row      : {'name': 'Ben', 'age': 27, 'field': 'Law'}
one single cell     : Ben



Read that output as a table: the **list** holds the rows in order, each **dictionary** is one row, and the keys (`name`, `age`, `field`) are the column names — which is why `participants[1]["name"]` picks out exactly one cell. This is not a toy analogy: a list of dictionaries is the shape data actually arrives in from surveys, databases and web APIs.

And it is precisely the shape **pandas** takes over in Notebook 5. There, the same rows become a `DataFrame` that you can inspect, filter, average per column and plot — without ever writing a position lookup by hand again.

### 💬 Discuss (60 seconds, with your neighbour)

Each person above is now three fields: `name`, `age`, `field`. What has been lost — and who decided which columns exist? When an AI system ranks job applicants or flags benefit claims, it sees rows like these and nothing else. Which column would you insist on adding, and which one would you refuse to record at all?


## 🧠 Key takeaways

1. **One value per variable does not scale.** Real data comes in collections, and Python's three everyday ones are the list, the tuple and the dictionary.
2. **List** — `[4, 5, 3]`: a series of similar things, looked up by **position** (`scores[0]`, `scores[-1]` for the last), measured with `len()`, changed item-by-item, grown with `.append()`. `scores[0:3]` slices out the first three.
3. **Tuple** — `(52.52, 13.40)`: a small fixed record. **Immutable** — assigning into it raises `TypeError` — and best used with **unpacking**: `lat, lon = location`.
4. **Dictionary** — `{"age": 31}`: one thing with labelled fields, looked up by **key** (`person["age"]`). The same syntax adds and updates; `in` tests membership; `.keys()` / `.values()` show labels and contents.
5. **`KeyError`** means "no such field". Use `person["age"]` when the key must exist and `person.get("age", fallback)` when it might not.
6. **Lists number their contents, dictionaries name them, tuples refuse to be edited.** Pick by the job, not by habit.
7. A **list of dictionaries is a table** — one dict per row, keys as column names. Exactly what pandas turns into a `DataFrame` in Notebook 5.

## ✅ Self-assessment

Tick what you can honestly check — anything shaky gets reinforced in the next notebooks:

- [ ] I can explain why a list beats 300 separate variables
- [ ] I can create a list and read its first and last item (and I know why the first is `[0]`)
- [ ] I can use `len()`, change an item, and `.append()` a new one
- [ ] I can say what "immutable" means and why a tuple is sometimes the *better* choice
- [ ] I can unpack a tuple with `lat, lon = location`
- [ ] I can create a dictionary, read a value by key, and add or update a field
- [ ] I can check with `in` whether a key exists, and avoid a `KeyError` with `.get()`
- [ ] I can look at a list of dictionaries and see a table

## 🚀 Next step

Continue with **Notebook 3 — Control Flow** ([`03_control_flow.ipynb`](03_control_flow.ipynb)): teaching Python to make **decisions** (`if / elif / else`) and to **repeat** work for you (`for`, `while`) — which is how you finally get to *walk through* one of today's collections instead of reading it item by item.

> **Want to go deeper after the session?** The self-paced track in [`../02_advanced_self_learning/`](../02_advanced_self_learning/) revisits all of this in depth — [`03_lists_data_structures.ipynb`](../02_advanced_self_learning/03_lists_data_structures.ipynb) for sequences and slicing, [`04_dictionaries_advanced.ipynb`](../02_advanced_self_learning/04_dictionaries_advanced.ipynb) for nested data.